lib/jobs/ucs-discovery.js

```diff
@@ -79,9 +79,19 @@ function UcsDiscoveryJobFactory(
         
         return self.getRoot()
             .then(function(root) {
-                return [ root, self.createRackmounts(root), self.createServers(root) ];
+                return [ self.createRackmounts(root), self.createServers(root) ];
             })
-            .spread(function() {
+            .spread(function(rackmounts, servers) {
+                self.context.physicalNodeList = (self.context.physicalNodeList || []).concat(
+                    _.map(rackmounts, function(it) {
+                        return _.get(it, 'id');
+                    })
+                );
+                self.context.physicalNodeList = _.flattenDeep(self.context.physicalNodeList.concat(
+                    _.map(servers, function(it) {
+                        return [_.get(it, 'id'), _.get(it.relations[0], 'targets') ];
```

Comment:
There are some hidden rules in your design:
- There is at least 1 relation
- The relation type happens to be the "enclosure" relation
- The newly created "enclosure" relation is happen to be the first relation.

If the node relation is empty, then you will get a "empty" node in final `self.context.physicalNodeList`, but you forgot to remove the empty node.

If the node relation is not "enclosure" relation or the "enclosure" relation is not at the first position, that means you are trying to handle wrong node, this will make  the ucs catalog fail.

These kinds of hidden rules make the code error-prone, especially somebody tries to refactor the code that generating the ucs server node but he isn't aware such kind of hidden rules. So my suggestion is:
- Filter the relation target by relation type, not use the fixed position.
- Always check whether the relation is empty or not.